### ISSUE TITLE
Problem to connect at Metalnx-web using SSL

### ISSUE THREAD

**Issue by felipegutierrez** [Metalnx/metalnx-web#17]
Dear,

I am trying to connect at my Metalnx-web new installation through SSL and it is not working. Could you help me please?
I installed the latest rpm that I found at https://bintray.com/metalnx/rpm/emc-metalnx-web .
I run the script `python /opt/emc/setup_metalnx.py` and it cannot connect to iRODS. So I commented the lines `subprocess.check_call(irods_auth_params, stdout=os_devnull)` to connect to iRODS in the file `lib/utils.py`.
The war generated by this script didn't deploy well. So I download the project from the github and compiled it, changing the configurations at the files `database.properties`, `irods.environment.properties`, `mysql.properties` and `security.properties`.
I am using the URL https://145.117.144.231:8443/emc-metalnx-web/login/ . When I type the user `rods` and its password I see on the logs that Metalnx-web is still using HTTP.
```
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "POST /emc-metalnx-web/login/authenticate/ HTTP/1.1" 302 -
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/login/serverNotResponding/ HTTP/1.1" 200 4563
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/css/bootstrap.min.css HTTP/1.1" 200 109587
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/images/logo.svg HTTP/1.1" 200 10686
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/css/font-awesome.min.css HTTP/1.1" 200 27466
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/css/styles.css HTTP/1.1" 200 30188
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/js/jquery.min.js HTTP/1.1" 200 95786
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/css/bootstrapValidator.min.css HTTP/1.1" 200 565
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/js/bootstrapValidator.min.js HTTP/1.1" 200 103245
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/css/bootstrapValidator.min.css HTTP/1.1" 200 565
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/js/bootstrapValidator.min.js HTTP/1.1" 200 103245
felipes-mbp.amc.nl - - [18/Apr/2017:11:00:06 -0400] "GET /emc-metalnx-web/fonts/Lato-Regular.ttf HTTP/1.1" 200 120196
```
Maybe I am missing something to configure. Do you have any clue of how can I fix it?

Kind Regards, Felipe

**Comment by ajguerra** [comment]
What do your config files look like?

**Comment by felipegutierrez** [comment]
File `security.properties` only the line `encoded.properties=`
File `mysql.properties` >
```
db.driverClassName=com.mysql.jdbc.Driver
db.url=jdbc:mysql://145.117.144.231:3306/metalnx
db.username=metalnx
db.password=metalnx
hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
hibernate.show_sql=false
hibernate.format_sql=false
hibernate.hbm2ddl.auto=update
connection.pool_size=10
```
File `irods.environment.properties` >
```
irods.host=145.117.144.231
irods.port=1247
irods.zoneName=tempZone
irods.auth.scheme=STANDARD
jobs.irods.username=rods
jobs.irods.password=123456
jobs.irods.auth.scheme=STANDARD
runSyncJobs=true
rmd.connection.timeout=500
rmd.connection.port=8000
reverse.dns.lookup=false
```
File `database.properties` >
```
db.driverClassName=com.mysql.jdbc.Driver
db.url=jdbc:mysql://145.117.144.231:3306/metalnx
db.username=metalnx
db.password=metalnx
hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
hibernate.show_sql=false
hibernate.format_sql=false
hibernate.hbm2ddl.auto=update
connection.pool_size=10
```

**Comment by ajguerra** [comment]
Your config files look fine... Can you post the beginning of your core.re file as well?

**Comment by felipegutierrez** [comment]
the only property I changed on the `core.re` file is `acPreConnect(*OUT) { *OUT="CS_NEG_REQUIRE"; }` to use SSL in iCommands. Which is working on the server and on a remote client.

**Comment by ajguerra** [comment]
I apologize. The core.re file isn't in play here. I misunderstood the problem. Let me take a quick look at the code here.

**Comment by neogeo68** [comment]
I'm not sure if it is related, but I made some changes in methods that change the database... Maybe if you drop Metalnx db and create it again might solve the problem.

**Comment by ajguerra** [comment]
I assume that you are using PAM auth in iRODS, right @felipegutierrez ?

**Comment by felipegutierrez** [comment]
Hi @neogeo68 . I drop the database and created it again with `setup_metalnx.py`. I noticed that when I set the configurations of iRODS the scripts does not ask about SSL certificates.
```
iRODS Paramaters:
    * irods.port = 1247
    * irods.auth.scheme = STANDARD
    * irods.zoneName = tempZone
    * jobs.irods.username = rods
    * irods.host = 145.117.144.231
```
On my `/var/lib/irods/.irods/irods_environment.json` file I have these parameters to use SSL certificate.
```
{
    "irods_host": "irodsprodvm.ebioscience.amc.nl",
    "irods_port": 1247,
    "irods_default_resource": "demoResc",
    "irods_home": "/tempZone/home/rods",
    "irods_cwd": "/tempZone/home/rods",
    "irods_user_name": "rods",
    "irods_zone_name": "tempZone",
    "irods_client_server_negotiation": "request_server_negotiation",
    "irods_client_server_policy": "CS_NEG_REQUIRE",
    "irods_encryption_key_size": 32,
    "irods_encryption_salt_size": 8,
    "irods_encryption_num_hash_rounds": 16,
    "irods_encryption_algorithm": "AES-256-CBC",
    "irods_default_hash_scheme": "SHA256",
    "irods_match_hash_policy": "compatible",
    "irods_server_control_plane_port": 1248,
    "irods_server_control_plane_key": "TEMPORARY__32byte_ctrl_plane_key",
    "irods_server_control_plane_encryption_num_hash_rounds": 16,
    "irods_server_control_plane_encryption_algorithm": "AES-256-CBC",
    "irods_maximum_size_for_single_buffer_in_megabytes": 32,
    "irods_default_number_of_transfer_threads": 4,
    "irods_transfer_buffer_size_for_parallel_transfer_in_megabytes": 4,
    "irods_ssl_certificate_chain_file": "/etc/irods/ssl/irods.crt",
    "irods_ssl_certificate_key_file": "/etc/irods/ssl/irods.key",
    "irods_ssl_dh_params_file": "/etc/irods/ssl/dhparams.pem",
    "irods_ssl_ca_certificate_file": "/etc/irods/ssl/irods.crt"
}
```
Hi @ajguerra . No. I am only using SSL, but no PAM anymore.

**Comment by ajguerra** [comment]
@felipegutierrez I posted an article on our Wiki about how to handle SSL in iRODS and Metalnx. I also added some information about PAM based on your first issue you opened here. 

The article has 4 sections: SSL config in iRODs, PAM, Metalnx and Tomcat. Since you mentioned that you have an iRODS grid already set up with SSL, I think you can skip the first two sections and go straight to the Metalnx and Tomcat configuration parts. 

There is an explaination about how you tell the setup_script to look for the certificates as you mentioned. It may not be the best solution yet, but it is working. We can talk about adding some more intelligence to it later on if you want.

At any rate, take a look at the [Wiki post](https://github.com/Metalnx/metalnx-web/wiki/PAM-&-SSL-Configuration) when you can. Let me know how things go.

Cheers!

**Comment by felipegutierrez** [comment]
many thanks. I will do =)

**Comment by felipegutierrez** [comment]
Hi @ajguerra , I got to put SSL to work with Metalnx through your Wiki post. THanks very much =)

By now I am getting the message `You do not have permissions to access information in the target collection or for the selected object.` when I try to search for collection of all users (rods and normal user). I am gonna put on the dev environment to debug the code. Maybe it is the same problem of PostgreSQL x MySQL. If this is the problem I fork again the new version of Metalnx-web and create the queries and services like you mentioned before.

by the way, are you from Campinas, Brasil? Because I saw on the SSL certificate.

**Comment by sgworth** [comment]
Felipe,
Jumping in as Arthur is away at present.  Looking into the source code this error message happens only if you try to look at the object into tab (in the Collections view) on a file or object you have no permission to.  The other place would be if you try to enter a location on the breadcrumb you do not have access to.    Can you provide details on the steps you performed so we can try to minic what you see on our systems?

**Comment by ajguerra** [comment]
Hi @felipegutierrez ,

Did you figure out whether or not the problem was related to the sql queries? I think it's a good idea for you to try using your queries for a MySQL DB.

**Comment by felipegutierrez** [comment]
hi, sorry for the late. I will try to figure out what is. The problem is that I have to update my irule that it is the priority. But after that Metalnx is on the queue. Lasts days I look at the Tomcat log and irods log and I didn't see any thing strange. So it is why I think I need to debug. But when I have some answer I will communicate here. Best Regards

**Comment by ajguerra** [comment]
What do you mean by updating your irule?

**Comment by felipegutierrez** [comment]
our production environment. it is not related to Metalnx-web

**Comment by ajguerra** [comment]
So, is everything working now?

**Comment by felipegutierrez** [comment]
Hi @ajguerra , I changed the attributes `hibernate.show_sql=true`, `log4j.rootLogger=DEBUG, stdout`, `log4j.logger.org.springframework=DEBUG`, `log4j.logger.org.thymeleaf=DEBUG`, `log4j.logger.org.irods=DEBUG`, `log4j.logger.org.hibernate.SQL=DEBUG`, `log4j.logger.org.hibernate.type=DEBUG` but I cannot see from where comes the message `You do not have permissions to access information in the target collection or for the selected object.      Back`.

I am checking the logs from tomcat `/usr/share/tomcat/logs`.

Do you have any clue?
Thanks

**Comment by ajguerra** [comment]
@felipegutierrez I can see this message in two places: `collectionInfo.html` and `collectionsBrowser.html` (this text is replaced on the span tag with `th:text="#{collections.management.msg.access.denied}"`).

`collectionInfo.html` is returned from the server when a request to `/info/` is made (check out `CollectionController`).

`collectionsBrowser.html` I think is more related to your problem because it is returned by the server everytime you browse through collections. So, take a look again at `CollectionController`, more specifically at the `getSubdirectories` and `getCollBrowserView` methods. This last one actually retrieves permissions for the user on the current path, so the problem might be there.

Let me know if it helped.

**Comment by ajguerra** [comment]
@felipegutierrez did you get your stuff working?

**Comment by felipegutierrez** [comment]
Hi @ajguerra ,
I think I have a problem with SSL on my iRODS server. Because I can deploy the same Metalnx-web.war on an iRODS server with no SSL and it works. So the queries are fine. On my iRODS server with SSL I can connect using iCommands and also using Metalnx-web, but I still cannot search for collections. If I change the IP to search for Collections on the remote MySQL with no SSL it works. It is strange, because I access the Metalnx-web using SSL [https://145.117.144.231:8443/emc-metalnx-web/collections/](https://145.117.144.231:8443/emc-metalnx-web/collections/) and use the MySQL host to the server with no SSL and it access. When I change back to my localhost MySQL it does not show the collections. And does not show SQL errors.
First I was thinking the problem was Grant privileges and I fixed it. Both servers have the same grants but only the one that does not use SSL can list the COllections.

**Comment by felipegutierrez** [comment]
I have some clue. This error appears on iRODS server only when I login Metalnx with SSL. Therefore I still can login, but I cannot list collections.
```
Jun  7 09:14:56 pid:13061 NOTICE: Agent exiting with status = 0
Jun  7 09:14:56 pid:13060 ERROR: [-]	iRODS/server/re/src/rules.cpp:670:actionTableLookUp :  status [PLUGIN_ERROR_MISSING_SHARED_OBJECT]  errno [] -- message []
	[-]	iRODS/server/re/src/irods_ms_plugin.cpp:110:load_microservice_plugin :  status [PLUGIN_ERROR_MISSING_SHARED_OBJECT]  errno [] -- message [Failed to create ms plugin entry.]
		[-]	iRODS/lib/core/include/irods_load_plugin.hpp:175:load_plugin :  status [PLUGIN_ERROR_MISSING_SHARED_OBJECT]  errno [] -- message [shared library does not exist [/var/lib/irods/plugins/microservices/libmsiobjget_microservices.so]]

Jun  7 09:14:56 pid:13060 ERROR: executeRuleAction Failed for msiobjget_microservices status = -1097000 NO_RULE_OR_MSI_FUNCTION_FOUND_ERR
```

**Comment by ajguerra** [comment]
@felipegutierrez those errors you see when you login showed up because you either have an old version of the Metalnx MSI package or don't have it installed at all. The `msiobjget_microservices` in particular is used on the dashboard page to list the microservices you have available in iRODS. So, I don't think this is related to the collections page issue. 

Back to the problem... I just put up an iRODS instance with SSL turned on. Everything worked fine in Metalnx. I also set up Metalnx with HTTP and HTTPS and in both cases the app was able to show items on the collections page. If you are in a dev environment it might be worth setting up a Postgres instance just to confirm everything works fine on your end too.

What I can tell you is that SSL has been a real issue for us. Jargon actually released two other versions trying to fix those SSL and performance issues we found (by *we* I include you as well because you found the first issues on PAM authentication). 

Some of those issues could be fixed in Jargon, but some couldn't since they were iRODS bugs. I'm not fully aware of all of them, but the Jargon devs run into issues with the iRODS API, the consortium's confirmed that issues do exist and they will be fixed in the next version (I think 4.1.11). Perharps your bug may be part of it (if it's not something else).
